java/maven/src/org/netbeans/modules/maven/problems/MavenModelProblemsProvider.java

```diff
@@ -133,70 +151,108 @@ public Collection<? extends ProjectProblem> getProblems() {
                 assert problemReporter != null;
                 project.getLookup().lookup(NbMavenProject.class).addPropertyChangeListener(projectListener);
             
             }
-            
+            MavenProject o = analysedProject.get();
+            LOG.log(Level.FINER, "Called getProblems for {0}, analysed = {1}, current = {2}", 
+                    new Object[] { project, o == null ? 0 : System.identityHashCode(o), System.identityHashCode(updatedPrj) });
             //for non changed project models, no need to recalculate, always return the cached value
-            Object wasprocessed = prj.getContextValue(MavenModelProblemsProvider.class.getName());
-            if (wasprocessed != null) {
-                Collection<ProjectProblem> cached = problemsCache.get();
-                LOG.log(Level.FINER, "Project was processed, cached is: {0}", cached);
+            Object wasprocessed = updatedPrj.getContextValue(MavenModelProblemsProvider.class.getName());
+            if (o == updatedPrj && wasprocessed != null) {
+                Pair<Collection<ProjectProblem>, Boolean> cached = problemsCache.get();
+                LOG.log(Level.FINER, "getProblems: Project was processed, cached is: {0}", cached);
                 if (cached != null) {
                     return cached;
                 }
             } 
-            Callable<Collection<? extends ProjectProblem>> c = new Callable<Collection<? extends ProjectProblem>>() {
-                @Override
-                public Collection<? extends ProjectProblem> call() throws Exception {
-                    Object wasprocessed = prj.getContextValue(MavenModelProblemsProvider.class.getName());
-                    if (wasprocessed != null) {
-                        Collection<ProjectProblem> cached = problemsCache.get();
-                        LOG.log(Level.FINER, "Project was processed #2, cached is: {0}", cached);
-                        if (cached != null) {                            
-                            return cached;
+            
+            SanityBuildAction sba = cachedSanityBuild.get();
+            if (sba != null && sba.getPendingResult() == null) {
+                cachedSanityBuild.clear();
+            }
+            c = () -> {
+                // double check, the project may be invalidated during the time.
+                MavenProject prj = ((NbMavenProjectImpl)project).getFreshOriginalMavenProject();
+                Object wasprocessed2 = updatedPrj.getContextValue(MavenModelProblemsProvider.class.getName());
+                if (o == updatedPrj && wasprocessed2 != null) {
+                    Pair<Collection<ProjectProblem>, Boolean> cached = problemsCache.get();
+                    LOG.log(Level.FINER, "getProblems: Project was processed #2, cached is: {0}", cached);
+                    if (cached != null) {                            
+                        return cached;
+                    }
```

Comment:
if you synchronize this bit, the `problemsCache` can become a regular reference.

Reply:
@sdedic this snippet isn't synchronized since it is in the callable block which can run in a different thread.

Reply:
> @sdedic this snippet isn't synchronized since it is in the callable block which can run in a different thread.

Uh-oh. Thanks.